lib/listen/record.rb

```diff
@@ -32,27 +27,20 @@ def unset_path(dir, rel_path)
     def file_data(dir, rel_path)
       root = @paths[dir.to_s]
       dirname, basename = Pathname(rel_path).split.map(&:to_s)
-      if [nil, '', '.'].include? dirname
-        root[basename] ||= {}
-        root[basename].dup
-      else
-        root[dirname] ||= {}
-        root[dirname][basename] ||= {}
-        root[dirname][basename].dup
-      end
+      dirname = '.' if [nil, '', '.'].include? dirname
+      root[dirname] ||= {}
+      root[dirname][basename] ||= {}
+      root[dirname][basename].dup
     end
 
     def dir_entries(dir, rel_path)
-      tree = if [nil, '', '.'].include? rel_path.to_s
-               @paths[dir.to_s]
-             else
-               @paths[dir.to_s][rel_path.to_s] ||= _auto_hash
-               @paths[dir.to_s][rel_path.to_s]
-             end
+      rel_path = '.' if [nil, '', '.'].include? rel_path.to_s
+      @paths[dir.to_s][rel_path.to_s] ||= _auto_hash
+      tree = @paths[dir.to_s][rel_path.to_s]
 
       result = {}
       tree.each do |key, values|
-        # only get data for file entries
+        # only get data for file/dir entries
         result[key] = values.key?(:mtime) ? values : {}
```

Comment:
I don't understand the separation of thought on '.' and other folders. Surely they should work in exactly the same way. That is, '.' is a path and in that path we have files [x,y,z] and '.git' is a path and in that path we have files [a,b,c]. This is what my patch did, unified them. All logic branch separating were removed and the only logic left was to make [nil,'','.'] the same path.

When I mentioned mtime I meant the key. At the moment the list is of files only so returning tree directly is exactly the same as it is now because all entries, being files, contain mtime anyway - the logic here does nothing. It still needs to be made though so add_dir adds the directory to the list of entries for its parent folder though as I mentioned at L74, and it can be an empty hash, which then WOULD mean that mtime key is only present for files.

I agree there needs to be acceptance test first though. I might take a look how they work a bit more (there's lots of undocumented helpers!) so I can offer more help if you need any.